### PR TITLE
Update CxxWrap to 0.9.0

### PR DESCRIPTION
**Pull request by SylvainCorlay**

### CODE REVIEW

**Comment by SylvainCorlay** [comment]
It seems that `mapped_julia_type` has been removed. Will need to study the new API to make that update.

**Comment by barche** [comment]
The easiest way to find out where this is coming from would be to run the test in the debugger, setting a breakpoint on the line where the `No appropriate factory for type` is thrown in libcxxwrap and then look at the backtrace. I tried doing that myself, but when trying to compile the tests on Arch Linux I get:

```
[cmake] CMake Error in test/CMakeLists.txt:
[cmake]   No known features for CXX compiler
[cmake] 
[cmake]   ""
[cmake] 
[cmake]   version .
[cmake] 
[cmake] 
```

**Comment by SylvainCorlay** [comment]
In `No appropriate factory for type i` /  `No appropriate factory for type d`, `i` and `d` appear to stand respectivelt for int and double. These are the element types of my arrays.

**Comment by SylvainCorlay** [comment]
I don't know about the cmake issue you are seeing. I presume this may come from the way I enable the C++17 as the CXX_STANDARD in cmake which may not work of old versions.

**Comment by barche** [comment]
Yes, but these standard types should of course be mapped automatically, so I need to determine the exact code path taken to get to that error, since it should be "caught" by the automatic type creation, which I made more explicit in the recent update, but I probably forgot a case.

**Comment by barche** [comment]
On cmake 3.15.4 here.

**Comment by SylvainCorlay** [comment]
> On cmake 3.15.4 here.

Same :/

**Comment by barche** [comment]
Looking over the code, it seems the CxxWrap Julia module is not loaded for these tests, that would explain it, since it fills a map with the types on module init.

**Comment by SylvainCorlay** [comment]
Thanks @barche!

The tests for xtensor-julia are driven from the C++ side (while Xtensor.jl is tested from the Julia side). I tried calling `jlcxx::initialize()` from the main.cpp but this appears to segfault unfortunately.

**Comment by barche** [comment]
OK, I will look into making some intialization function that is callable from C++. The technical reason for all this is that I found static variables in class templates to be fragile on Windows, since they are not unique between DLL boundaries, so there is now a global type cache in a std::map, and this needs to be initialized for the fundamental types.

**Comment by SylvainCorlay** [comment]
Okay thanks for the heads up. Initializing from C++ would be great!

This will also be useful for xeus-julia which will embed the julia interpreter as well.

**Comment by barche** [comment]
Just pushed a change to libcxxwrap-julia master that should allow the `register_core_types` to work (or at least not segfault ;)

**Comment by SylvainCorlay** [comment]
Thanks for looking into this!

I restarted the build to pick up the latest master (on Travis) and I am still seeing the segfault.

**Comment by barche** [comment]
Ah, yes, this doesn't pick up libcxxwrap master, unfortunately. I could do another point release, but it would be more reassuring if you could build and test locally first.

**Comment by SylvainCorlay** [comment]
I just tried with libcxxwrap-julia master and it still fails, but only with one type:

```
C++ exception with description "Type m has no Julia wrapper" thrown in the test body.
```

Build log is available here: https://travis-ci.org/xtensor-stack/xtensor-julia/jobs/613233582?utm_medium=notification&utm_source=github_status

**Comment by SylvainCorlay** [comment]
@barche - this appears to be `unsigned long`. libcxxwrap-julia may be missing some basic types.

**Comment by barche** [comment]
Unfortunately `unsigned long` has to be in CxxWrap and can't be linked to an existing integer type in Julia base, because it might be the same as `Int`, or not, depending on the platform. In the update to 0.9, I made sure every distinct C++ integer type has a distinct Julia equivalent, which wasn't the case before and strictly speaking it must be done to avoid the possibility of ambiguous methods.

https://en.cppreference.com/w/cpp/language/types

So that type will only work with the CxxWrap module loaded and initialized.

**Comment by SylvainCorlay** [comment]
> Unfortunately unsigned long has to be in CxxWrap and can't be linked to an existing integer type in Julia base, because it might be the same as Int, or not, depending on the platform. In the update to 0.9, I made sure every distinct C++ integer type has a distinct Julia equivalent, which wasn't the case before and strictly speaking it must be done to avoid the possibility of ambiguous methods.
> 
> https://en.cppreference.com/w/cpp/language/types
>
> So that type will only work with the CxxWrap module loaded and initialized.

Is it an issue that the type mapping it is not a bijective mapping. Why should it not be possible to bind `unsigned long` to Int even if `Int` corresponds to other C++ types?

Preventing this makes it much harder to *embed* Julia with libcxxwrap, which is also the approach we have for other projects.

**Comment by SylvainCorlay** [comment]
@barche so it seems that your change to make `register_core_types` callable from C++ worked.

Now the remaining issue is that I cannot expose `unsigned long` integers to Julia when embedding Julia in my application.

**Comment by barche** [comment]
I know, still had this in my backlog to respond :)

What I plan to try is to make it easy using the libcxxwrap API to load and initialize the CxxWrap module from C++, I think this will be the most robust solution.

The reason the mapping needs to be bijective is that it is otherwise impossible to distinguish between e.g. `std::vector<unsigned long>` and `std::vector<uint64_t>`.

**Comment by SylvainCorlay** [comment]
> What I plan to try is to make it easy using the libcxxwrap API to load and initialize the CxxWrap module from C++, I think this will be the most robust solution.

Sounds a bit convoluted though?

> The reason the mapping needs to be bijective is that it is otherwise impossible to distinguish between e.g. `std::vector<unsigned long>` and `std::vector<uint64_t>`.

In which situation do we want to make that distinction from Julia?

**Comment by barche** [comment]
> > What I plan to try is to make it easy using the libcxxwrap API to load and initialize the CxxWrap module from C++, I think this will be the most robust solution.
> 
> Sounds a bit convoluted though?

Yes, but I think in the end it may be the only way, since also the `CxxRef` and `CxxPtr`, which are used to encapsulate any pointer type, are defined on the CxxWrap Julia side. Initialization is happening anyway when doing `jl_init`, so loading the CxxWrap module should not add that much extra complexity. One point to becareful about is the Julia environment, probably it's best to provide a way to choose which one to use.

> > The reason the mapping needs to be bijective is that it is otherwise impossible to distinguish between e.g. `std::vector<unsigned long>` and `std::vector<uint64_t>`.
> 
> In which situation do we want to make that distinction from Julia?

It's not really necessary for Julia, but all methods are bound to a type. I'm not sure it's OK to call e.g.  the C++ function `std::vector<unsigned long>::push_back` on an object that is of type `std::vector<uint64_t>` (on platforms where these types are not the same).

**Comment by barche** [comment]
OK, in the latest libcxxwrap and CxxWrap master versions, everything should work. You should replace the `jl_init()` calls with `jlcxx::cxxwrap_init();`, and make sure in the test script that the CxxWrap module is installed (already the case I think?).

**Comment by SylvainCorlay** [comment]
> OK, in the latest libcxxwrap and CxxWrap master versions, everything should work. You should replace the jl_init() calls with jlcxx::cxxwrap_init();, and make sure in the test script that the CxxWrap module is installed (already the case I think?).

Hum, actually it is not... I need to set up the installation of CxxWrap to use the general prefix installation of libcxxwrap...

**Comment by barche** [comment]
In addition to the typo, I think the travis file is still missing a step like this, to be run from the libcxxwrap build directory after building libcxxwrap:

```- JLCXX_DIR=$PWD julia -e 'using Pkg; pkg"add CxxWrap#master"; Pkg.build(); using CxxWrap'```

**Comment by SylvainCorlay** [comment]
> `JLCXX_DIR=$PWD julia -e 'using Pkg; pkg"add CxxWrap#master"; Pkg.build(); using CxxWrap'`.

Although it still seems weird to me that we need the julia package for the embedding case :/

**Comment by barche** [comment]
I think it is inevitable, in the new CxxWrap we also have `CxxRef`, `CxxConstRef`, `CxxPtr` and `CxxConstPtr` defined on the Julia side, so sooner or later you would probably hit those, too.

**Comment by SylvainCorlay** [comment]
Still getting a `Error initializing CxxWrap module`...

File: .travis.yml [hunk 1]
```diff
@@ -60,14 +60,17 @@ install:
...
-    # Retrieve JlCxx_DIR from CxxWrap
-    - JlCxx_DIR=$(julia -E "using CxxWrap; joinpath(dirname(pathof(CxxWrap)), \"..\", \"deps\", \"usr\", \"lib\", \"cmake\", \"JlCxx\")")
-    - JlCxx_DIR=${JlCxx_DIR//\"/}
+    - git clone https://github.com/JuliaInterop/libcxxwrap-julia.git
+    - cd libcxxwrap-julia
+    - mkdir build; cd build
+    - cmake -D CMAKE_INSTALL_PREFIX=$HOME/miniconda ..
+    - make install
+    - cd ../../
+    - JLCXX_DIR=$CMAKE_INSTALL_PREFIX julia -E "using Pkg; Pkg.add(PackageSpec(name=\"CxxWrap\", rev=\"master\"))"
```
Comment (barche) [inline .travis.yml]: Is it possible `$CMAKE_INSTALL_PREFIX` is empty?

Comment (SylvainCorlay) [reply .travis.yml]: good catch! I this is fixed now - although the build is still failing.

**Comment by SylvainCorlay** [comment]
```bash
test_xtensor_julia: /home/travis/julia/include/julia/julia.h:800: jl_value_t* jl_svecset(void*, size_t, void*):
Assertion `jl_typeis(t,jl_simplevector_type)' failed.
```

**Comment by barche** [comment]
The good news is that I can reproduce this on my machine. The bad news is that I have no idea why this happens. This line is the culprit, but `jl_svecset` is used in exactly the same way in Julia itself as far as I can tell:

https://github.com/JuliaInterop/libcxxwrap-julia/blob/61f3acc2cf7dd577c5509ab586fd928805518aa5/include/jlcxx/module.hpp#L366

**Comment by barche** [comment]
On a new try, it seems random, can't reproduce it anymore and all tests pass locally on my machine.

**Comment by SylvainCorlay** [comment]
> On a new try, it seems random, can't reproduce it anymore and all tests pass locally on my machine.

If it's random, then we are fine ;)

**Comment by barche** [comment]
I think it is related to a valgrind error in `jl_init`, which seems to occur only on linux. I need to investigate this further, but maybe it's a good idea to get the macOS builds working in the mean time, that seems to be a problem with compiler version?

**Comment by barche** [comment]
I think the problem was the improper initialization for the STL module, this should be fixed in master. Fingers crossed :)

**Comment by barche** [comment]
That's green on Linux, let's test it 1000 more times to be sure ;)

**Comment by SylvainCorlay** [comment]
> That's green on Linux, let's test it 1000 more times to be sure ;)

Yes! This is great news.

**Comment by SylvainCorlay** [comment]
Hey @barche 
Since you released libcxxwrap-julia 0.6.6 and CxxWrap 0.9.0, I looked into finishing this up.

- This now works on OS X and Linux (dropping compilers that don't support C++17), but it still fails on Windows, at compile time, with mingw.
- It seems to fail on fold expressions with Visual Studio 2015 and 2017.

**Comment by barche** [comment]
I have only tested with MSVC2019 on Windows. The MinGW seems to be because the GCC is too old, it needs at least GCC7 in the current state.

**Comment by SylvainCorlay** [comment]
I am working on a PR to make CxxWrap work with C++14.

Conda-forge needs either MSVC 2015 or the mingw in question.

The previous version of CxxWrap worked with GCC 4.9, MSVC 2015, Clang 3.9... It will be a long time before we can require such recent versions...

**Comment by SylvainCorlay** [comment]
Closing as fixed in master.

**Pull request closed without merging**